lib/phoenix_live_view/upload_config.ex

```diff
@@ -0,0 +1,570 @@
+defmodule Phoenix.LiveView.UploadEntry do
+  @moduledoc """
+  The struct representing an upload entry.
+  """
+
+  alias Phoenix.LiveView.UploadEntry
+
+  defstruct progress: 0,
+            upload_config: nil,
+            upload_ref: nil,
+            ref: nil,
+            valid?: false,
+            done?: false,
+            cancelled?: false,
+            client_name: nil,
+            client_size: nil,
+            client_type: nil,
+            client_last_modified: nil
+
+  @type t :: %__MODULE__{
+          progress: integer(),
+          upload_config: String.t() | :atom,
+          upload_ref: String.t(),
+          ref: String.t() | nil,
+          valid?: boolean(),
+          done?: boolean(),
+          cancelled?: boolean(),
+          client_name: String.t() | nil,
+          client_size: integer() | nil,
+          client_type: String.t() | nil,
+          client_last_modified: integer() | nil
+        }
+
+  @doc false
+  def put_progress(%UploadEntry{} = entry, 100) do
+    %UploadEntry{entry | progress: 100, done?: true}
+  end
+
+  def put_progress(%UploadEntry{} = entry, progress) do
+    %UploadEntry{entry | progress: progress}
+  end
+end
+
+defmodule Phoenix.LiveView.UploadConfig do
+  @moduledoc """
+  The struct representing an upload.
+  """
+
+  alias Phoenix.LiveView.UploadConfig
+  alias Phoenix.LiveView.UploadEntry
+
+  @default_max_file_size 8_000_000
+  @default_chunk_size 64_000
+  @default_chunk_timeout 10_000
+
+  @unregistered :unregistered
+  @invalid :invalid
+
+  @too_many_files :too_many_files
+
+  defstruct name: nil,
+            client_key: nil,
+            max_entries: 1,
+            max_file_size: @default_max_file_size,
+            chunk_size: @default_chunk_size,
+            chunk_timeout: @default_chunk_timeout,
+            entries: [],
+            entry_refs_to_pids: %{},
+            entry_refs_to_metas: %{},
+            accept: %{},
+            external: false,
+            allowed?: false,
+            ref: nil,
+            errors: []
```

Comment:
Are all of these fields public? Perhaps we should `@derive Inspect` with only the fields the user should be accessing?